### ISSUE TITLE
accuracy sacrificed with multi-gpu using gluon

### ISSUE THREAD

**Issue by GarrickLin** [apache/incubator-mxnet#7615]
## Environment info
Operating System:
Ubuntu 16.04

MXNet version: 
0.11.0

Or if installed from source:
no, using pip install mxnet-cu80

Python version and distribution:
Python27

I run the demo in [`P06-C03-object-detection.ipynb`](https://github.com/zackchase/mxnet-the-straight-dope/blob/master/P06-C03-object-detection.ipynb)

and try to modify to use multi-gpu version as follows
```
    for epoch in range(start_epoch, epochs):
        train_data.reset()
        cls_metric.reset()
        box_metric.reset()
        tic = time.time()
        # iterate through all batch
        for i, batch in enumerate(train_data):
            btic = time.time()
            # record gradients:
            with ag.record():
                if isinstance(ctx, mx.context.Context):
                    X = [batch.data[0].as_in_context(ctx)]
                    Y = [batch.label[0].as_in_context(ctx)]
                elif isinstance(ctx, list):
                    X = gluon.utils.split_and_load(batch.data[0], ctx)
                    Y = gluon.utils.split_and_load(batch.label[0], ctx)
                else:
                    raise Exception("Unexpected context type")
                loss = []
                for x, y in zip(X, Y):
                    default_anchors, class_predictions, box_predictions = net(x)
                    box_target, box_mask, cls_target = training_targets(default_anchors, class_predictions, y)
                    # losses
                    loss1 = cls_loss(class_predictions, cls_target)
                    loss2 = box_loss(box_predictions, box_target, box_mask)
                    # sum all losses
                    loss.append(loss1 + loss2)
                for l in loss:
                    # backpropagate
                    l.backward()
            # apply
            trainer.step(batch_size)
            # update metrics
            cls_metric.update([cls_target], [nd.transpose(class_predictions, (0, 2, 1))])
            box_metric.update([box_target], [box_predictions * box_mask])
            if (i + 1) % log_interval == 0:
                name1, val1 = cls_metric.get()
                name2, val2 = box_metric.get()
                print '[Epoch %d Batch %d] speed: %f samples/s, training: %s=%f, %s=%f' % \
                      (epoch, i, batch_size / (time.time() - btic), name1, val1, name2, val2)
        nd.waitall()
        # end of epoch logging
        name1, val1 = cls_metric.get()
        name2, val2 = box_metric.get()
        print '[Epoch %d] training: %s=%f, %s=%f' % (epoch, name1, val1, name2, val2)
        print '[Epoch %d] time cost: %f' % (epoch, time.time() - tic)
```
and I try to train single-gpu and multi-gpu respectively
here are the results after 165 epochs from scratch

1. single-gpu
![](http://ww1.sinaimg.cn/mw690/9ddd8b3bly1fiw9iwppcsj20rs0rsapr.jpg)

2. multi-gpu
![](http://ww1.sinaimg.cn/mw690/9ddd8b3bly1fiw9kgp9loj20rs0rs7k0.jpg)

Is that resonable? I am not familiar with new grammar and not sure wether I made a mistale.

**Comment by GarrickLin** [comment]
for 4 gpu training batchsize is 256
for 1 gpu training batchsize is 32

**Comment by piiswrong** [comment]
Probably because you changed batchsize but didn't adjust learning rate?

@zhreshold

**Comment by zhreshold** [comment]
@GarrickLin So you manually modified the batchsize right?

**Comment by GarrickLin** [comment]
yes

**Comment by zhreshold** [comment]
Try to reduce the batch-size, since there's a hidden mini-batch in the algorithm itself.

**Comment by GarrickLin** [comment]
I get the proper result after set batch_size back to 32, but the speed is obviously slowing down even worse than single-gpu. How can I adjust learning rate to suit the 256 batchsize for 4 gpu training?

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
